src/varyDSL.js

```diff
@@ -1,125 +1,107 @@
 import TestTrackConfig from './testTrackConfig';
 
-var VaryDSL = function(options) {
-  if (!options.assignment) {
-    throw new Error('must provide assignment');
-  } else if (!options.visitor) {
-    throw new Error('must provide visitor');
-  }
-
-  this._assignment = options.assignment;
-  this._visitor = options.visitor;
-  this._splitRegistry = TestTrackConfig.getSplitRegistry();
-
-  this._variantHandlers = {};
-};
-
-VaryDSL.prototype.when = function() {
-  // these 5 lines are messy because they ensure that we throw the most appropriate error message if when is called incorrectly.
-  // the benefit of this complexity is exercised in the test suite.
-  var argArray = Array.prototype.slice.call(arguments, 0),
-    lastIndex = argArray.length - 1,
-    firstArgIsVariant = typeof argArray[0] !== 'function' && argArray.length > 0,
-    variants = firstArgIsVariant ? argArray.slice(0, Math.max(1, lastIndex)) : [],
-    handler = argArray[lastIndex];
-
-  if (variants.length === 0) {
-    throw new Error('must provide at least one variant');
-  }
+class VaryDSL {
+  constructor(options) {
+    if (!options.assignment) {
+      throw new Error('must provide assignment');
+    } else if (!options.visitor) {
+      throw new Error('must provide visitor');
+    }
 
-  for (var i = 0; i < variants.length; i++) {
-    this._assignHandlerToVariant(variants[i], handler);
-  }
-};
+    this._assignment = options.assignment;
+    this._visitor = options.visitor;
+    this._splitRegistry = TestTrackConfig.getSplitRegistry();
 
-VaryDSL.prototype.default = function(variant, handler) {
-  if (this._defaultVariant) {
-    throw new Error('must provide exactly one `default`');
+    this._variantHandlers = {};
   }
 
-  this._defaultVariant = this._assignHandlerToVariant(variant, handler);
-};
+  when(...args) {
+    const handler = typeof args[args.length - 1] === 'function' ? args.pop() : null;
```

Comment:
👍 this took a sec to noodle through, but looks good.

is there a downside to mutating `args` by calling `pop`? if someone were to pass in their own array, would it be modified or does the runtime make a copy?

Reply:
The way the ["rest" arguments](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/rest_parameters#Description) works, it converts the list of arguments passed into the function as a new array. If someone were to pass an array as an argument to this function it would appear as just an item in the `...args` array. 

```js
when(['earth', 'wind', 'fire', handlerFunction]);

when(...args) {
  console.log(args); // [['earth', 'wind', 'fire', handlerFunction]]
}
```

So to answer the question "if someone were to pass in their own array, would it be modified", then no because `args` is a brand new array that contains that array passed in.

Reply:
Would be nice to put the variants in the tail position. Then we can have `when(handler, ...variants)`